dlt/destinations/impl/clickhouse/clickhouse.py

```diff
@@ -0,0 +1,334 @@
+
+                table_function = (
+                    f"azureBlobStorage('{storage_account_url}','{container_name}','{ blobpath }','{ account_name }','{ account_key }','{ clickhouse_format}')"
+                )
+
+            else:
+                # Unsigned access.
+                raise NotImplementedError(
+                    "Unsigned Azure Blob Storage access from Clickhouse isn't supported as yet."
+                )
+
+        with client.begin_transaction():
+            client.execute_sql(
+                f"""INSERT INTO {qualified_table_name} SELECT * FROM {table_function}"""
+            )
+
+    def state(self) -> TLoadJobState:
+        return "completed"
+
+    def exception(self) -> str:
+        raise NotImplementedError()
+
+
+class ClickhouseClient(SqlJobClientWithStaging, SupportsStagingDestination):
+    capabilities: ClassVar[DestinationCapabilitiesContext] = capabilities()
+
+    def __init__(
+        self,
+        schema: Schema,
+        config: ClickhouseClientConfiguration,
+    ) -> None:
+        self.sql_client: ClickhouseSqlClient = ClickhouseSqlClient(
+            config.normalize_dataset_name(schema), config.credentials
+        )
+        super().__init__(schema, config, self.sql_client)
+        self.config: ClickhouseClientConfiguration = config
+        self.active_hints = deepcopy(HINT_TO_CLICKHOUSE_ATTR)
+        self.type_mapper = ClickhouseTypeMapper(self.capabilities)
+
+    def start_file_load(self, table: TTableSchema, file_path: str, load_id: str) -> LoadJob:
+        return super().start_file_load(table, file_path, load_id) or ClickhouseLoadJob(
+            file_path,
+            table["name"],
+            self.sql_client,
+            staging_credentials=(
+                self.config.staging_config.credentials if self.config.staging_config else None
+            ),
+        )
+
+    def _get_table_update_sql(
+        self, table_name: str, new_columns: Sequence[TColumnSchema], generate_alter: bool
+    ) -> List[str]:
+        table: TTableSchema = self.prepare_load_table(table_name, self.in_staging_mode)
+        sql = SqlJobClientBase._get_table_update_sql(self, table_name, new_columns, generate_alter)
+
+        if generate_alter:
+            return sql
+
+        # Default to 'ReplicatedMergeTree' if user didn't explicitly set a table engine hint.
+        table_type = cast(
+            TTableEngineType, table.get(TABLE_ENGINE_TYPE_HINT, "replicated_merge_tree")
+        )
+        sql[0] = f"{sql[0]}\nENGINE = {TABLE_ENGINE_TYPE_TO_CLICKHOUSE_ATTR.get(table_type)}"
+
+        if primary_key_list := [
+            self.capabilities.escape_identifier(c["name"])
+            for c in new_columns
+            if c.get("primary_key")
+        ]:
+            sql[0] += "\nPRIMARY KEY (" + ", ".join(primary_key_list) + ")"
+        else:
+            sql[0] += "\nPRIMARY KEY tuple()"
+
+        # TODO: Apply sort order and cluster key hints.
```

Comment:
also for follow up ticket

Reply:
https://github.com/dlt-hub/dlt/issues/1183